Add Mining Levels and Tool Attributes API

NoComment1105 opened the pull request:
This pull request ports the Mining Level API and the Tool Attributes API from the Fabric API to Quilt Standard Libararies.
In its current state it is a near direct port, with only minor changes to some class names, due to other QSL naming.
I am opening this pull-request as a draft for now, because i wish to rewrite certain parts of this, I am aware other wish to do this too, but I don't want to do anything bad, so it's here for people to let me know when I've done somethings wrong

NoComment1105 committed:
Mining Levels and Tool Attributes

NoComment1105 committed:
Fix some licenseing in package-info.java and some javadoc

OroArmor commented on library/core/qsl_base/src/main/java/org/quiltmc/qsl/base/api/util/BooleanFunction.java:
```diff
@@ -0,0 +1,34 @@
+ * <p>This is the {@code boolean}-consuming primitive specialization for {@link java.util.function.Function}.
+ */
+@FunctionalInterface
+public interface BooleanFunction<R> {
```
Idea: extend `Function<Boolean, R>`

NoComment1105 replied:
done

OroArmor commented on library/item/tool_attributes/src/main/java/org/quiltmc/qsl/tool_attributes/api/QuiltToolTags.java:
```diff
@@ -0,0 +1,39 @@
+import net.minecraft.util.Identifier;
+import org.quiltmc.qsl.tag.api.TagRegistry;
+
+public class QuiltToolTags {
```
Javadocs probably wanted

NoComment1105 replied:
Added

OroArmor reviewed (approved):
Looks very good so far.

NoComment1105 committed:
Add suggestions

NoComment1105 committed:
Add BooleanFunction.java suggestion from @Kroppeb

NoComment1105 committed:
Merge branch 'QuiltMC:1.18' into mining-and-tools

NoComment1105 committed:
Merge branch 'QuiltMC:1.18' into mining-and-tools

NoComment1105 committed:
Start update to 22w03a

NoComment1105 committed:
Updates and fixes

NoComment1105 committed:
Merge branch 'QuiltMC:1.18' into mining-and-tools

NoComment1105 committed:
Remove deprecated methods

Remove un-needed package specifications from javadocs

Fix some spacing issues

silence intellij errors

NoComment1105 committed:
Certified Genius:tm:

NoComment1105 committed:
Fix some more errors and add `interLibraryDependencies` method to build script

NoComment1105 committed:
Minor testmod changes

OroArmor committed:
temp

NoComment1105 committed:
Remove some un-needed entries in json file

EnnuiL commented:
It has been, like, more than a year since this PR was opened; I'll be closing this
